### ISSUE TITLE
Enable Lifecycle Events to be customizable (2)

### ISSUE THREAD

**Issue by CypressJosh** [CypressJosh/jira-demo#50]
## Why this is necessary

Cypress automatically executes code in "between" tests. The idea behind this is that Cypress will try to "reset the state" and enable each test to run in its own pure state without other tests affecting or leaking into one another.

Currently it does the following:
- clears `localStorage`
- clears `sessionStorage`
- clears `cookies`
- removes aliases you've created
- resets routes you've defined
- removes stubs, spies, clocks you've created
- clears the cache once before the browser opens

Currently what it **does not** do:
- clears the actual application under test

## Why this needs to change

It really makes no sense the way we currently handle this. This is often a source of confusion we've seen from multiple users.

For instance it is:

- not clear that this is happening
- not clear how you can stop lifecycle events from applying
- not clear why we reset everything above but not the application itself (?)

Because we don't forcibly clear the application state it means that it "appears" nothing has changed between the tests, but under the hood everything that built up the current session was reset.

This often results with users clicking links only to find out they've been magically "logged out".

In addition to sessions being cleared, it is actually technically possible for your application to do things in the tiniest briefest moment that Cypress finishes running one test and "switches" to the new test. Since your application is still active in the window, that means it will continue to run everything it's built to run.

This can cause all kinds of havoc.

Here's a sequence of events that demonstrates the problem:

- before test 1
  - cypress resets state asynchronously
- test 1 begins
  - your cypress spec code runs
  - you visit your application
  - you setup `cy.route` to route XHR's
  - cypress finishes executing all commands
- test 1 finishes
  - cypress moves onto test 2
- before test 2
  - cypress resets state asynchronously
     - while this happens, the application you visited in test 1 runs a poll / timer which creates an XHR
     - the state from the routes has been reset, so instead of XHR being routed, it goes out to your server
     - your server throws a 500, and the `onerror` handler of your XHR bubbles up and changes your application's state
      - your application is now in an invalid state and you have no idea why
- test 2 begins
  - your cypress spec code runs
  - you don't visit your application again (you already did in test 1)
  - your commands fail because the state of your application in test 1 is now invalid
 - test 2 finishes, displays error

The reason we don't hear **every** user complain about this - is that for one its incredibly rare for this to happen and represents a "perfect storm" of events.

Additionally when we originally built these lifecycle events, they were with the assumption that most users would be visiting their applications **before each test**.

When visiting before each test (and thus always recreating / rebuilding the sessions and state), you insulate yourself from all of these problems.

## What users really want to do

With that said many of our users really want to do one thing: login once in a `before` hook during the first test, and then have all subsequent tests run with this session.

While we don't really officially endorse this as the best pattern - it often makes sense in these situations:

- your app loads slow (albeit you have a much bigger problem than writing tests)
- you're not actually mutating session state in tests and therefore theres no reason to enforce needing to have it cleared each time

Another example users have suggested is when you apply a "unit testing" pattern to integration tests and write a bunch of "tiny" tests with a single assertion. However, we consider this an anti-pattern.

Irrespective of what we believe - the bottom line is that only logging in once before your tests will without a doubt enable all of your tests to run faster. If they have less to do in each test, they will simply complete faster.

Therefore it's imperative that we adequately solve this use case, and in the documentation and guides we provide resources / suggestions / hints / tips / tricks / philosophy to guide you to making these choices.

BTW in case you're curious we believe most of the need for a single login in a `before` hook is mitigated completely by using programatic API's like `cy.request` for things like setting state and logging in. By avoiding your UI you skip most of the reasons tests are slow.

One thing we absolutely **do not endorse** and will never support is writing tests

#### Clear the App

Enforce clearing the Application at the end of test *if it is not the last test* but *before the next test runs any of its code*.

In other words, unless this is the very last test, the moment test code finishes and before any async code runs we MUST blow the application out of the window so no async events, polls, timers, promises, etc, run.

This will 100% prevent application leakage in "between" the tests and ensure the application is torn down properly.

#### Update the GUI

Display a new "instrument section" that lists each lifecycle event that was applied (or skipped).

![screen shot 2017-09-23 at 8 03 26 pm](https://user-images.githubusercontent.com/1268976/30778089-5d1f23f0-a09a-11e7-9162-e38940f7ac65.png)

#### Create new commnad

We'll need a new `cy.clearApp()` command which brings into parity the other `cy.clear*` commands.

This enables you to issue it within a test as a "one-off" command the same way the others work.

#### Remove old API's

Deprecate / remove the `Cypress.Cookies.preserveOnce` API's since that would be superseded by these Lifecycle events.

## Other Considerations

Having first class lifecycle events with their own programatic API also means we could turn these completely off when writing **unit tests**. As it stands Cypress automatically runs the clearing code in between each test. These are asynchronous and involve talking to the automation servers (in the browser and server) and therefore are quite slow (in terms of clock cycles).

## Examples of how we could do this

I don't like the idea of making this global configuration in `cypress.json`. Instead I think we should create API's that enable you to control this in code.

Since Cypress has to know how you intend to apply lifecycle events this code would have to go **before** any tests are defined or run.

Likely this is what you want - you'd likely want these changes to propagate to an entire spec file.

```js
// users_spec.js

Cypress.lifecycle() // returns an object with the current configuration

Cypress.lifecycle(false) // disable all lifecycle events
Cypress.lifecycle(true) // enable all lifecycle events

Cypress.lifecycle({
  clearApp: true, // leave this on
  clearInternals: true // leave this on
  clearCookies: false // nope
  clearLocalStorage: false // nope
  clearSessionStorage: false // nope
})

describe('my suite', () => {
  it('test 1', () => {
    ...
  })

  it('test 2', () => {
    ...
  })
})
```

Users wanting to apply Lifecycle events globally to all specs would just add the code into a support file.

## Use Cases

At first glance it seems as if lifecycle events would be binary: you'd either want to have them all enabled (by default) or turn them all off.

Not so fast Jack - let's take a deeper look and why there's a bit more complexity than this:

### Clear Everything (the default)

**Pros**
- Absolutely no state build up
- Simple, clear, insulated
- Promotes taking shortcuts and skipping the UI as much as possible

**Cons**
- Realistically a bit slower to run all tests

**Summary**
- Slowest but flake free. Easy to understand.

### Clear Nothing (what many users want)

**Pros**
- You visit once in a `before` hook, usually establishing the session there
- All subsequent tests use this existing session and don't visit
- The tests will run faster

**Cons**
- All of the tests are loosely coupled to one another unless you're really careful
- The state from one test will **immediately** apply to the subsequent test
- When running a single test vs all you will likely see **different** behavior that is difficult to debug / understand
- You're still susceptible to actions in your application leaking into other tests (as described above in detail)

**Summary**
- Fastest but susceptible to edge cases, potential flake, and poorly written tests.

### Clear Some Things (likely what you should do instead)

If you want the benefit of only establishing a session once, but don't want to problems of loosely coupled tests then you likely want this configuration:

```js
Cypress.lifecycle(false) // disable everything
Cypress.lifecycle{
  clearApp: true // but always clear the app
})

describe('my suite', function () {
  before(function () {
    cy.login() // establish the session only once!
  })

  beforeEach(function () {
    cy.visit('/') // but continue to visit before each test
  })

  it('test 1')

  it('test 2')

  it('test 3')
})
```

**Pros**
- The current state of the app is always blown away before each test
- The session is preserved between tests avoiding this to be re-created

**Cons**
- Takes more explanation and deeper understanding of Cypress to know about this
- Requires visiting before each test since the app is blown away
- Keeping the same session around can lead to state issues on the server (if the server isn't stubbed)
- You're apt not to take shortcuts which means that the gains you receive from running all tests are lost when running and iterating on a single test

**Summary**
- Faster than the first open, slower than 2nd. Likely what you want instead.
